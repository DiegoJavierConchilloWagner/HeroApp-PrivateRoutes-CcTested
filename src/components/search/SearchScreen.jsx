import React, { useMemo } from 'react'
import queryString from 'query-string';
import { useLocation } from 'react-router';

import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHerosByName';

export const SearchScreen = ({ history }) => {

    const location = useLocation();
    const { q = '' } = (queryString.parse( location.search ));

    const initialForm = { searchText: q };
    const [ formValues, handleInputChange, ] = useForm( initialForm );
    const { searchText } = formValues;
    
    const heroesFiltered = useMemo(() => getHeroesByName( q ), [q])

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${ searchText }`)
    };

    return (
        <div>
            <h1>SearchScreen</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4> Search Form</h4>
                    <hr />

                    <form onSubmit={ handleSearch }>
                        <input type="text" name="searchText" value={searchText} onChange={ handleInputChange } placeholder="Find your Hero" className="form-control"/>
                        <button type="submit" className="btn m-1 btn-block btn-primary"> 
                            Search...
                        </button>
                    </form>

                </div>

                <div className="col-7">
                    <h4> Results </h4>
                    <hr />
                    {   
                        (q === '') &&
                            <div className="alert alert-info">
                                Search a hero
                            </div>
                    }
                    {   
                        (q !== '' && heroesFiltered.length === 0 ) &&
                            <div className="alert alert-danger">
                                There is no a hero with { q }
                            </div>
                    }
                    {
                        heroesFiltered.map( hero => (
                            <HeroCard key={ hero.id } { ...hero } />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}
