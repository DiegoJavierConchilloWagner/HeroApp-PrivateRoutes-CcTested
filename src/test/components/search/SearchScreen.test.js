import React from 'react';
import { mount } from 'enzyme'
import { MemoryRouter, Route } from 'react-router';
import { SearchScreen } from '../../../components/search/SearchScreen';



describe('Pruebas en el componente SearchScreen', () => {
    
    const wrapper = mount(
        <MemoryRouter initialEntries={['/search']}>
            <Route path='/search' component={SearchScreen} />
        </MemoryRouter>
    )

    test('Debe de mostrarse correctamente con valores por defecto', () => {

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-info').text() ).toBe('Search a hero')

    });
    
    test('Debe de mostrar a Bartman y el input con el valor del QueryString', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path='/search' component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( wrapper.find('input').prop('value') ).toBe('batman');
        
    });
    
    test('Debe de mostrar un error si no se encuentra el Hero', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=512623']}>
                <Route path='/search' component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( wrapper.find('.alert-danger').text() )
            .toBe('There is no a hero with 512623');
        
    });

    test('Debe de llamar el push del history', () => {

        const history = {
            push: jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <Route path='/search' component={ 
                    () => < SearchScreen history={ history } /> } />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change',{
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect( history.push ).toHaveBeenCalledWith(`?q=batman`);
        
    });
});
