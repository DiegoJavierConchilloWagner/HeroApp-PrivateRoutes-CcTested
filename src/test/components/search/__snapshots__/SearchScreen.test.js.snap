// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente SearchScreen Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
